File: leetcode/2466. Count Ways To Build Good Strings/main.cc
#include <unordered_map>

const int MOD = 1e9 + 7;

class Solution {
public:
  static int countGoodStrings(int low, int high, int zero, int one) {
    std::unordered_map<int, int>
        dp;
    return dfs(0, low, high, zero, one, dp);
  }

private:
  static int dfs(int length, int low, int high, int zero, int one,
                 std::unordered_map<int, int> &dp) {
    if (length > high) {
      return 0;
    }

    if (dp.find(length) != dp.end()) {
      return dp[length];
    }

    int count = (length >= low) ? 1 : 0;
    count += dfs(length + zero, low, high, zero, one, dp);
    count %= MOD;
    count += dfs(length + one, low, high, zero, one, dp);
    count %= MOD;

    dp[length] = count;

    return count;
  }
};

File: leetcode/2466. Count Ways To Build Good Strings/main.rs
use std::collections::HashMap;

const MOD: i32 = 1_000_000_007;

impl Solution {
    pub fn count_good_strings(low: i32, high: i32, zero: i32, one: i32) -> i32 {
        let mut dp: HashMap<i32, i32> = HashMap::new();

        Self::dfs(0, low, high, zero, one, &mut dp)
    }

    fn dfs(
        length: i32,
        low: i32,
        high: i32,
        zero: i32,
        one: i32,
        dp: &mut HashMap<i32, i32>,
    ) -> i32 {
        if length > high {
            return 0;
        }

        if let Some(&val) = dp.get(&length) {
            return val;
        }

        let mut res = if length >= low { 1 } else { 0 };

        res += Self::dfs(length + zero, low, high, zero, one, dp);
        res %= MOD;
        res += Self::dfs(length + one, low, high, zero, one, dp);
        res %= MOD;

        dp.insert(length, res);

        res
    }
}
